File: 2021/08/03/linked_list_intersection.py
"""
https://leetcode.com/problems/intersection-of-two-linked-lists/
Author: Andrew Jarombek
Date: 8/3/2021
"""


from typing import Optional


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


def get_intersection_node(head_a: ListNode, head_b: ListNode) -> Optional[ListNode]:
    nodes_a = set()

    while head_a is not None:
        nodes_a.add(id(head_a))
        head_a = head_a.next

    while head_b is not None:
        if id(head_b) in nodes_a:
            return head_b

        head_b = head_b.next

    return None

File: 2022/03/14/c/linkedListIntersection.c
// Program: https://leetcode.com/problems/intersection-of-two-linked-lists/
// Author: Andrew Jarombek
// Date: 3/14/2022

#include <stdlib.h>

struct ListNode {
    int val;
    struct ListNode* next;
};

struct ListNode* getIntersectionNode(struct ListNode* headA, struct ListNode* headB) {
    int aLength = 0;
    struct ListNode *nodeA = headA;
    struct ListNode *nodeB = headB;

    while (nodeA != NULL) {
        aLength++;
        nodeA = nodeA->next;
    }

    int bLength = 0;

    while (nodeB != NULL) {
        bLength++;
        nodeB = nodeB->next;
    }

    for (int i = aLength - bLength; i > 0; i--) {
        headA = headA->next;
    }

    for (int i = bLength - aLength; i > 0; i--) {
        headB = headB->next;
    }

    while (headA != NULL) {
        if (headA == headB) {
            return headA;
        }

        headA = headA->next;
        headB = headB->next;
    }

    return NULL;
}

int main() {}
